jembi/JeMPI
# Updating the matching/linking rules syntax to support null, and repeated variable names

**Ticket:** N/A
**Other Related Tickets:** N/A

----

## Describe of changes

This change extends the config rules parser to support `null` values, and have multiple variables in the rull. The rules that can now be supported are examplified below:

- Example1
   - Input rule: `eq(nationalId)`
   - Output
     ```graphql
     query query_link_deterministic_a($national_id: string) {
            all(func:type(GoldenRecord)) @filter(eq(GoldenRecord.national_id, 
              $national_id
              )) {
               uid
               GoldenRecord.source_id {
                  uid
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
            }
         }
      ``` 

- Example2
   - Input rule: `eq(national_id) or null(national_id) and eq(ccc)`
   - Output
     ```graphql
     query query_link_deterministic_a($national_id: string) {
            var(func:type(GoldenRecord)) @filter(eq(GoldenRecord.national_id, 
              $national_id
              )) {
               A as uid
            }
            var(func:type(GoldenRecord)) @filter(eq(GoldenRecord.national_id, "")) {
               B as uid
            }
            var(func:type(GoldenRecord)) @filter(eq(GoldenRecord.ccc, 
              $ccc
              )) {
               C as uid
            }
            all(func:type(GoldenRecord)) @filter(uid(A) OR (uid(B) AND uid(C))) {
               uid
               GoldenRecord.source_id {
                  uid
               }
               GoldenRecord.aux_date_created
               GoldenRecord.aux_auto_update_enabled
               GoldenRecord.aux_id
               GoldenRecord.given_name
               GoldenRecord.family_name
               GoldenRecord.gender
               GoldenRecord.dob
               GoldenRecord.city
               GoldenRecord.phone_number
               GoldenRecord.national_id
            }
         }
      ``` 

## How to test


- To test this you will need to adjist the config reference ( JeMPI_Apps/JeMPI_Configuration/reference/config-reference-api.json ) rules to the `null` expression. An example of this if is found in the following file [config-reference.json](https://github.com/jembi/JeMPI/files/14420278/config-reference.json)
- Once update you will need the source code generator (run `./create.sh reference/config-reference.json` in the 'JeMPI_Apps/JeMPI_Configuration/reference/' folder)
- From there you should be able to add new golden records, and search as appropiate to get your results back. An example of addin records (based on the config above) and search is found in the thunderclient config 
[JeMPI-QueryTests.json](https://github.com/jembi/JeMPI/files/14420455/JeMPI-QueryTests.json)
